fuse/ipns/ipns_unix.go

```diff
@@ -192,115 +191,78 @@ func (r *Root) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
-	// NOTE: this strategy wont work well if we allow hard links
-	// (im all for murdering the thought of hard links)
-	name string
+	fs.NodeRef
+}
 
-	// Private keys held by nodes at the root of a keyspace
-	// WARNING(security): the PrivKey interface is currently insecure
-	// (holds the raw key). It will be secured later.
-	key ci.PrivKey
+// File is wrapper over an ipnsfs file to satisfy the fuse fs interface
+type File struct {
+	fi *nsfs.File
 
-	Ipfs   *core.IpfsNode
-	Nd     *mdag.Node
-	dagMod *mod.DagModifier
-	cached *ftpb.Data
+	fs.NodeRef
 }
 
-func (s *Node) loadData() error {
-	s.cached = new(ftpb.Data)
-	return proto.Unmarshal(s.Nd.Data, s.cached)
+// Attr returns the attributes of a given node.
+func (d *Directory) Attr() fuse.Attr {
+	log.Debug("Directory Attr")
+	return fuse.Attr{Mode: os.ModeDir | 0555}
 }
 
 // Attr returns the attributes of a given node.
-func (s *Node) Attr() fuse.Attr {
-	if s.cached == nil {
-		err := s.loadData()
-		if err != nil {
-			log.Debugf("Error loading PBData for file: '%s'", s.name)
-		}
+func (fi *File) Attr() fuse.Attr {
+	log.Debug("File Attr")
+	size, err := fi.fi.Size()
+	if err != nil {
+		// In this case, the dag node in question may not be unixfs
+		log.Critical("Failed to get file size: %s", err)
 	}
-	switch s.cached.GetType() {
-	case ftpb.Data_Directory:
-		return fuse.Attr{Mode: os.ModeDir | 0555}
-	case ftpb.Data_File, ftpb.Data_Raw:
-		size, err := ft.DataSize(s.Nd.Data)
-		if err != nil {
-			log.Debugf("Error getting size of file: %s", err)
-			size = 0
-		}
-		if size == 0 {
-			dmsize, err := s.dagMod.Size()
-			if err != nil {
-				log.Error(err)
-			}
-			size = uint64(dmsize)
-		}
-
-		mode := os.FileMode(0666)
-		if IpnsReadonly {
-			mode = 0444
-		}
-
-		return fuse.Attr{
-			Mode:   mode,
-			Size:   size,
-			Blocks: uint64(len(s.Nd.Links)),
-		}
-	default:
-		log.Debug("Invalid data type.")
-		return fuse.Attr{}
+	return fuse.Attr{
+		Mode: os.FileMode(0666),
+		Size: uint64(size),
 	}
 }
 
 // Lookup performs a lookup under this node.
-func (s *Node) Lookup(ctx context.Context, name string) (fs.Node, error) {
-	nodes, err := s.Ipfs.Resolver.ResolveLinks(s.Nd, []string{name})
+func (s *Directory) Lookup(ctx context.Context, name string) (fs.Node, error) {
+	child, err := s.dir.Child(name)
 	if err != nil {
 		// todo: make this error more versatile.
 		return nil, fuse.ENOENT
 	}
 
-	return s.makeChild(name, nodes[len(nodes)-1]), nil
-}
-
-func (n *Node) makeChild(name string, node *mdag.Node) *Node {
-	child := &Node{
-		Ipfs:   n.Ipfs,
-		Nd:     node,
-		name:   name,
-		nsRoot: n.nsRoot,
-		parent: n,
-	}
-
-	// Always ensure that each child knows where the root is
-	if n.nsRoot == nil {
-		child.nsRoot = n
-	} else {
-		child.nsRoot = n.nsRoot
+	switch child := child.(type) {
+	case *nsfs.Directory:
+		return &Directory{dir: child}, nil
+	case *nsfs.File:
+		return &File{fi: child}, nil
+	default:
+		panic("system has proven to be insane")
```

Comment:
This isnt a unixfs type, this is an ipnsfs Type, I only ever create Directories and Files. Its panic worthy